# Problem Statement

    In the telecom industry, customers are able to choose from multiple service providers and actively switch from one operator to another. In this highly competitive market, the telecommunications industry experiences an average of 15-25% annual churn rate. Given the fact that it costs 5-10 times more to acquire a new customer than to retain an existing one, customer retention has now become even more important than customer acquisition.

    For many incumbent operators, retaining high profitable customers is the number one business goal. To reduce customer churn, telecom companies need to predict which customers are at high risk of churn. In this project, you will analyze customer-level data of a leading telecom firm, build predictive models to identify customers at high risk of churn.

    In this competition, your goal is to build a machine learning model that is able to predict churning customers based on the features provided for their usage.

In [1]:
### ignoring all the warning that might come
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Import the numpy and pandas package

import numpy as np
import pandas as pd

# Importing visualization packages
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# Importing packages for Linear Regression
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler, PolynomialFeatures

# Importing RFE 
from sklearn.feature_selection import RFE
from sklearn.linear_model import LinearRegression

# importing r2_scoe, vif ans stats model
from sklearn.metrics import r2_score
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.preprocessing import PowerTransformer

# importing models
from sklearn import linear_model
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error, mean_absolute_error

In [3]:
# Settings the number of rows to display in output to few all column data
# got below code from "https://stackoverflow.com/questions/58298963/jupyter-not-showing-whole-output" url
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

## Step 1: Data Understanding, Preparation, and Pre-Processing
Reading and understanding the data in the dataset

In [4]:
# Loading train data in telecomtrain_df dataframe
tltrain_df = pd.read_csv("train.csv")
# printing top 5 rows for understanding
tltrain_df.head()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0   0        109             0.0             0.0             0.0   
1   1        109             0.0             0.0             0.0   
2   2        109             0.0             0.0             0.0   
3   3        109             0.0             0.0             0.0   
4   4        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8   arpu_6  \
0            6/30/2014            7/31/2014            8/31/2014   31.277   
1            6/30/2014            7/31/2014            8/31/2014    0.000   
2            6/30/2014            7/31/2014            8/31/2014   60.806   
3            6/30/2014            7/31/2014            8/31/2014  156.362   
4            6/30/2014            7/31/2014            8/31/2014  240.708   

    arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  onnet_mou_8  offnet_mou_6  \
0   87.009    7.527        48.58       124.38         1.29         32.24   
1  122.787   42.953         0.00         0.00         0.00          0.00   
2  103.176    0.000         0.53        15.93         0.00         53.99   
3  205.260  111.095         7.26        16.01         0.00         68.76   
4  128.191  101.565        21.28         4.83         6.13         56.99   

   offnet_mou_7  offnet_mou_8  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
0         96.68          2.33           0.00            0.0            0.0   
1         25.99         30.89           0.00            0.0            0.0   
2         82.05          0.00           0.00            0.0            0.0   
3         78.48         50.23           0.00            0.0            0.0   
4         38.11          9.63          53.64            0.0            0.0   

   roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_6  \
0           0.00            0.0           0.00              2.23   
1           0.00            0.0           0.00              0.00   
2           0.00            0.0           0.00              0.53   
3           0.00            0.0           1.63              6.99   
4          15.73            0.0           0.00             10.16   

   loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2m_mou_6  loc_og_t2m_mou_7  \
0              0.00              0.28              5.29             16.04   
1              0.00              0.00              0.00              0.00   
2             12.98              0.00             24.11              0.00   
3              3.94              0.00             37.91             44.89   
4              4.83              6.13             36.74             19.88   

   loc_og_t2m_mou_8  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
0              2.33              0.00              0.00              0.00   
1              0.00              0.00              0.00              0.00   
2              0.00              0.00              0.00              0.00   
3             23.63              0.00              0.00              0.00   
4              4.61             11.99              1.23              5.01   

   loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_mou_6  \
0              0.00              0.00              0.00          7.53   
1              0.00             22.01             29.79          0.00   
2              2.14              0.00              0.00         24.64   
3              0.00              0.00              8.03         44.91   
4              0.00              9.85              0.00         58.91   

   loc_og_mou_7  loc_og_mou_8  std_og_t2t_mou_6  std_og_t2t_mou_7  \
0         16.04          2.61             46.34            124.38   
1          0.00          0.00              0.00              0.00   
2         12.98          0.00              0.00              2.94   
3         48.84         23.63              0.26             12.06   
4         25.94         15.76              0.00              0.00   

   std_og_t2t_mo

In [5]:
# loading the data in telecom_df dataframe
tltest_df = pd.read_csv('test.csv')
# printing top 5 rows for understanding
tltest_df.head()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0  69999        109             0.0             0.0             0.0   
1  70000        109             0.0             0.0             0.0   
2  70001        109             0.0             0.0             0.0   
3  70002        109             0.0             0.0             0.0   
4  70003        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8   arpu_6  \
0            6/30/2014            7/31/2014            8/31/2014   91.882   
1            6/30/2014            7/31/2014            8/31/2014  414.168   
2            6/30/2014            7/31/2014            8/31/2014  329.844   
3            6/30/2014            7/31/2014            8/31/2014   43.550   
4            6/30/2014            7/31/2014            8/31/2014  306.854   

    arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  onnet_mou_8  offnet_mou_6  \
0   65.330   64.445        31.78        20.23        23.11         60.16   
1  515.568  360.868        75.51        41.21        19.84        474.34   
2  434.884  746.239         7.54         7.86         8.40         16.98   
3  171.390   24.400         5.31         2.16         0.00         40.04   
4  406.289  413.329       450.93       609.03       700.68         60.94   

   offnet_mou_7  offnet_mou_8  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
0         32.16         34.83           0.00           0.00           0.00   
1        621.84        394.94           0.00           0.00           0.00   
2         45.81         45.04          22.81         103.38          26.08   
3        205.01         24.01           0.00           0.00           0.00   
4         23.84         74.16           0.00           0.00           0.00   

   roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_6  \
0           0.00           0.00           0.00             24.88   
1           0.00           0.00           0.00             75.51   
2          24.53          53.68          54.44              0.00   
3           0.00           0.00           0.00              5.31   
4           0.00           0.00           0.00              0.45   

   loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2m_mou_6  loc_og_t2m_mou_7  \
0             20.23             21.06             18.13             10.89   
1             41.21             19.84            473.61            598.08   
2              0.00              0.00              0.00              0.00   
3              0.00              0.00              2.94             98.61   
4              0.78             14.56              2.39              2.66   

   loc_og_t2m_mou_8  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
0              8.36              0.00             13.58              0.00   
1            377.26              0.73              0.00              0.00   
2              0.00              0.00              0.00              0.00   
3             20.51              0.00              0.00              2.35   
4             10.94              0.00              0.00              0.00   

   loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_mou_6  \
0               0.0              0.00              0.03         43.01   
1               0.0              0.00              0.00        549.86   
2               0.0              0.00              0.00          0.00   
3               0.0              6.18              0.00          8.26   
4               0.0              0.00              0.00          2.84   

   loc_og_mou_7  loc_og_mou_8  std_og_t2t_mou_6  std_og_t2t_mou_7  \
0         44.71         29.43              6.90              0.00   
1        639.29        397.11              0.00              0.00   
2          0.00          0.00              0.00              0.00   
3         98.61         22.86              0.00              2.16   
4          3.44         25.51            450.48            608.24   

 

In [6]:
# printing the shape of training data
print(f'Training dataset shape: {tltrain_df.shape}')
# printing the shape of test data
print(f'Testing dataset shape: {tltest_df.shape}')

Training dataset shape: (69999, 172)
Testing dataset shape: (30000, 171)


So from the above data we understand that we have around 69999 rows and 172 collumns in train data 
and 30000 rows and 171 collumns in test data

### Understnding Data Distribution

In this segment we try to understand how data is distributed that is into numerical and categorical data 

In [7]:
# feature type summary
tltrain_df.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Data columns (total 172 columns):
 #   Column                    Dtype  
---  ------                    -----  
 0   id                        int64  
 1   circle_id                 int64  
 2   loc_og_t2o_mou            float64
 3   std_og_t2o_mou            float64
 4   loc_ic_t2o_mou            float64
 5   last_date_of_month_6      object 
 6   last_date_of_month_7      object 
 7   last_date_of_month_8      object 
 8   arpu_6                    float64
 9   arpu_7                    float64
 10  arpu_8                    float64
 11  onnet_mou_6               float64
 12  onnet_mou_7               float64
 13  onnet_mou_8               float64
 14  offnet_mou_6              float64
 15  offnet_mou_7              float64
 16  offnet_mou_8              float64
 17  roam_ic_mou_6             float64
 18  roam_ic_mou_7             float64
 19  roam_ic_mou_8             float64
 20  roam_og_mou_6             f

In [8]:
tltrain_df.describe()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count  69999.000000    69999.0         69297.0         69297.0   
mean   34999.000000      109.0             0.0             0.0   
std    20207.115084        0.0             0.0             0.0   
min        0.000000      109.0             0.0             0.0   
25%    17499.500000      109.0             0.0             0.0   
50%    34999.000000      109.0             0.0             0.0   
75%    52498.500000      109.0             0.0             0.0   
max    69998.000000      109.0             0.0             0.0   

       loc_ic_t2o_mou        arpu_6        arpu_7        arpu_8   onnet_mou_6  \
count         69297.0  69999.000000  69999.000000  69999.000000  67231.000000   
mean              0.0    283.134365    278.185912    278.858826    133.153275   
std               0.0    334.213918    344.366927    351.924315    299.963093   
min               0.0  -2258.709000  -1289.715000   -945.808000      0.000000   
25%               0.0     93.581000     86.714000     84.095000      7.410000   
50%               0.0    197.484000    191.588000    192.234000     34.110000   
75%               0.0    370.791000    365.369500    369.909000    119.390000   
max               0.0  27731.088000  35145.834000  33543.624000   7376.710000   

        onnet_mou_7   onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
count  67312.000000  66296.000000  67231.000000  67312.000000  66296.000000   
mean     133.894438    132.978257    198.874771    197.153383    196.543577   
std      311.277193    311.896596    316.818355    322.482226    324.089234   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.675000      6.410000     34.860000     32.240000     31.575000   
50%       32.280000     32.100000     96.480000     91.885000     91.800000   
75%      115.837500    115.060000    232.990000    227.630000    229.345000   
max     8157.780000  10752.560000   8362.360000   7043.980000  14007.340000   

       roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  \
count   67231.000000   67312.000000   66296.000000   67231.000000   
mean        9.765435       7.014568       7.004892      14.186457   
std        57.374429      55.960985      53.408135      73.469261   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max      2850.980000    4155.830000    4169.810000    3775.110000   

       roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_6  loc_og_t2t_mou_7  \
count   67312.000000   66296.000000      67231.000000      67312.000000   
mean        9.842191       9.771783         46.904854         46.166503   
std        58.511894      64.618388        150.971758        154.739002   
min         0.000000       0.000000          0.000000          0.000000   
25%         0.000000       0.000000          1.660000          1.650000   
50%         0.000000       0.000000         11.910000         11.580000   
75%         0.000000       0.000000         40.740000         39.760000   
max      2812.040000    5337.040000       6431.330000       7400.660000   

       loc_og_t2t_mou_8  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
count      66296.000000      67231.000000      67312.000000      66296.000000   
mean          45.686109         93.238231         90.799240         91.121447   
std          153.716880        162.046699        153.852597        152.997805   
min            0.000000          0.000000          0.000000          0.000000   
25%            1.610000          9.920000         10.090000          9.830000   
50%           11.740000         41.030000         40.170000         40.350000   
75%           39.895000        110.430000        107.540000        109.245000   
max        10752.560000       4696

In [9]:
# feature type summary
tltest_df.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 171 columns):
 #   Column                    Dtype  
---  ------                    -----  
 0   id                        int64  
 1   circle_id                 int64  
 2   loc_og_t2o_mou            float64
 3   std_og_t2o_mou            float64
 4   loc_ic_t2o_mou            float64
 5   last_date_of_month_6      object 
 6   last_date_of_month_7      object 
 7   last_date_of_month_8      object 
 8   arpu_6                    float64
 9   arpu_7                    float64
 10  arpu_8                    float64
 11  onnet_mou_6               float64
 12  onnet_mou_7               float64
 13  onnet_mou_8               float64
 14  offnet_mou_6              float64
 15  offnet_mou_7              float64
 16  offnet_mou_8              float64
 17  roam_ic_mou_6             float64
 18  roam_ic_mou_7             float64
 19  roam_ic_mou_8             float64
 20  roam_og_mou_6             f

In [10]:
tltest_df.describe()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count  30000.000000    30000.0         29684.0         29684.0   
mean   84998.500000      109.0             0.0             0.0   
std     8660.398374        0.0             0.0             0.0   
min    69999.000000      109.0             0.0             0.0   
25%    77498.750000      109.0             0.0             0.0   
50%    84998.500000      109.0             0.0             0.0   
75%    92498.250000      109.0             0.0             0.0   
max    99998.000000      109.0             0.0             0.0   

       loc_ic_t2o_mou        arpu_6        arpu_7        arpu_8   onnet_mou_6  \
count         29684.0  30000.000000  30000.000000  30000.000000  28831.000000   
mean              0.0    282.644348    279.355021    279.845167    130.629693   
std               0.0    314.560212    323.205370    326.437029    290.677420   
min               0.0   -185.043000  -2014.045000   -146.859000      0.000000   
25%               0.0     92.907500     87.504750     84.180500      7.280000   
50%               0.0    198.227500    191.775000    191.690500     34.760000   
75%               0.0    371.969250    365.250750    367.804750    117.170000   
max               0.0  10955.066000  13224.119000  13505.996000   6944.360000   

        onnet_mou_7   onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
count  28828.000000  28325.000000  28831.000000  28828.000000  28325.000000   
mean     133.148632    133.111348    195.745473    196.792375    196.647889   
std      302.921820    301.951806    316.923836    333.628481    334.277643   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.660000      6.540000     34.385000     32.110000     31.740000   
50%       32.550000     33.060000     95.830000     91.435000     92.810000   
75%      115.210000    117.730000    229.230000    224.972500    225.760000   
max     7411.010000   6723.790000   6962.460000   9667.130000  13547.840000   

       roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  \
count   28831.000000   28828.000000   28325.000000   28831.000000   
mean       10.380431       7.465889       7.967267      13.269785   
std        99.973352     103.337675      94.622596      66.475975   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max     13724.380000   15371.040000   13095.360000    2165.440000   

       roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_6  loc_og_t2t_mou_7  \
count   28828.000000   28325.000000      28831.000000      28828.000000   
mean        9.763957      10.440251         47.557601         47.188688   
std        58.325466      64.933372        150.588656        156.664295   
min         0.000000       0.000000          0.000000          0.000000   
25%         0.000000       0.000000          1.650000          1.580000   
50%         0.000000       0.000000         11.930000         11.680000   
75%         0.000000       0.000000         41.410000         40.190000   
max      1878.860000    2860.510000       5878.530000       7163.210000   

       loc_og_t2t_mou_8  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
count      28325.000000      28831.000000      28828.000000      28325.000000   
mean          46.359887         93.584273         92.793180         93.238288   
std          145.087354        164.481651        165.674188        164.516683   
min            0.000000          0.000000          0.000000          0.000000   
25%            1.580000          9.770000          9.890000          9.740000   
50%           11.690000         41.010000         41.060000         40.440000   
75%           40.490000        110.330000        107.645000        108.680000   
max         6723.790000       4729

#### Observations
* From the above segment we understand that how there are more numerical type data in both train and test data and some categorical variables 
* We also see that training data has 172 collumns and test has 171 collumns so the target variable churn probability is dropped in test set.

### Categorizing the values 

In [11]:
#All numerical variables in training data are on numtrain
num_train = tltrain_df.columns[(tltrain_df.dtypes=="float64")|(tltrain_df.dtypes=="int64")].tolist()
num_train.remove('churn_probability')
print('Training dataset Numeric Column count: ', len(num_train))

#All numerical variables in training data are on numtrain
num_test = tltest_df.columns[(tltest_df.dtypes=="float64")|(tltest_df.dtypes=="int64")].tolist()
print('Testing dataset Numeric Column count: ' , len(num_test))


#All categorical var in training data are in cat_train and removing target var
cat_train = tltrain_df.select_dtypes(exclude=np.number).columns.tolist()
print('Training dataset Non-Numeric Column count: ' , len(cat_train))

#All categorical var in training data are in cat_train and removing target var
cat_test = tltest_df.select_dtypes(exclude=np.number).columns.tolist()
print('Testing dataset Numeric Column count: ' , len(cat_test))

Training dataset Numeric Column count:  162
Testing dataset Numeric Column count:  162
Training dataset Non-Numeric Column count:  9
Testing dataset Numeric Column count:  9


So in the above section both train and test data are catagorised to numerical and categorical data based all float and int data type variables are considered as numerical and object type considered as categorical data.

And also by removing the `churn_probability` in train_set we see that both test and train data have same diemensions

In [12]:
# Printing columns with above 99% of unique values as well as single unqiue value
for column in tltest_df.columns:
#     getting the count of unique values
    unique_count = len(tltrain_df[column].unique())
#     checking for unique value count more than 99% as well as single value
    if unique_count == 1 or ((unique_count/tltrain_df.shape[0])*100) >= 99 :
        print('{0:20}  {1}'.format(column, unique_count))

id                    69999
circle_id             1
last_date_of_month_6  1


we can observe that, 
> All values of <b>id</b> column are unique

> <b>circle_id</b> & <b>last_date_of_month_6</b> column has only one value in it

These columns are not useful for analysis, so lets drop them

In [13]:
#Droping above columns in both test and training dataset
tltrain_df = tltrain_df.drop(['id','circle_id', 'last_date_of_month_6'],axis=1)
tltest_df = tltest_df.drop(['id','circle_id', 'last_date_of_month_6'],axis=1)

print('{0:10}  {1}'.format('Training dataset shape post dropping columns', tltrain_df.shape[1]))
print('{0:10}  {1}'.format('Testing dataset shape post dropping columns', tltest_df.shape[1]))

Training dataset shape post dropping columns  169
Testing dataset shape post dropping columns  168


> We can see that initially <b>172</b> columns in training dataset as well as <b>171</b> in test dataset, post dropping the columns we can see <b>169</b> and <b>168</b> columns respectively

### Imputing missing values

Now lets check the missing values and imputing appropripate values

In [14]:
#Checking missing values in train dataset
for column in tltest_df.columns:
    missing_record_count = tltrain_df[column].isna().sum()
#     checking for unique value count more than 99% as well as single value
    if missing_record_count != 0 :
        print('{0:30}  {1}'.format(column, missing_record_count))

loc_og_t2o_mou                  702
std_og_t2o_mou                  702
loc_ic_t2o_mou                  702
last_date_of_month_7            399
last_date_of_month_8            733
onnet_mou_6                     2768
onnet_mou_7                     2687
onnet_mou_8                     3703
offnet_mou_6                    2768
offnet_mou_7                    2687
offnet_mou_8                    3703
roam_ic_mou_6                   2768
roam_ic_mou_7                   2687
roam_ic_mou_8                   3703
roam_og_mou_6                   2768
roam_og_mou_7                   2687
roam_og_mou_8                   3703
loc_og_t2t_mou_6                2768
loc_og_t2t_mou_7                2687
loc_og_t2t_mou_8                3703
loc_og_t2m_mou_6                2768
loc_og_t2m_mou_7                2687
loc_og_t2m_mou_8                3703
loc_og_t2f_mou_6                2768
loc_og_t2f_mou_7                2687
loc_og_t2f_mou_8                3703
loc_og_t2c_mou_6                2768
loc_og

In [15]:
# Lets print number of coulms with missing values
print('Number of columns with missing values: ', len(tltrain_df.columns[tltrain_df.isna().any()].tolist()))

Number of columns with missing values:  125


<b>Now lets impute the missing values</b>

In [16]:
tltrain_df['loc_og_t2o_mou'].value_counts()

0.0    69297
Name: loc_og_t2o_mou, dtype: int64

> There is only 1 value in `loc_og_t2o_mou` column, which is not useful for analysis and can be dropped

In [17]:
tltrain_df['std_og_t2o_mou'].value_counts()

0.0    69297
Name: std_og_t2o_mou, dtype: int64

> There is only 1 value in `std_og_t2o_mou` column, which is not useful for analysis and can be dropped

In [18]:
tltrain_df['loc_ic_t2o_mou'].value_counts()

0.0    69297
Name: loc_ic_t2o_mou, dtype: int64

> There is only 1 value in `loc_ic_t2o_mou` column, which is not useful for analysis and can be dropped

In [19]:
tltrain_df['last_date_of_month_7'].value_counts()

7/31/2014    69600
Name: last_date_of_month_7, dtype: int64

In [20]:
tltrain_df['last_date_of_month_8'].value_counts()

8/31/2014    69266
Name: last_date_of_month_8, dtype: int64

In [21]:
tltrain_df['onnet_mou_6'].describe()

count    67231.000000
mean       133.153275
std        299.963093
min          0.000000
25%          7.410000
50%         34.110000
75%        119.390000
max       7376.710000
Name: onnet_mou_6, dtype: float64

`Minutes of usages of all calls within same network operator on 6th month` the missing value in the call means the customer is not called anyone, which means it's `0`

In [22]:
# imputing the onnet_mou_6 column with 0.0
tltrain_df['onnet_mou_6'] = tltrain_df['onnet_mou_6'].fillna(0.0)

In [23]:
tltrain_df['onnet_mou_7'].describe()

count    67312.000000
mean       133.894438
std        311.277193
min          0.000000
25%          6.675000
50%         32.280000
75%        115.837500
max       8157.780000
Name: onnet_mou_7, dtype: float64

`Minutes of usages of all calls within same network operator on 7th month` the missing value in the call means the customer is not called anyone, which means it's `0`

In [24]:
# imputing the onnet_mou_7 column with 0.0
tltrain_df['onnet_mou_7'] = tltrain_df['onnet_mou_7'].fillna(0.0)

In [25]:
tltrain_df['onnet_mou_8'].describe()

count    66296.000000
mean       132.978257
std        311.896596
min          0.000000
25%          6.410000
50%         32.100000
75%        115.060000
max      10752.560000
Name: onnet_mou_8, dtype: float64

`Minutes of usages of all calls within same network operator on 8th month` the missing value in the call means the customer is not called anyone, which means it's `0`

In [26]:
# imputing the onnet_mou_8 column with 0.0
tltrain_df['onnet_mou_8'] = tltrain_df['onnet_mou_8'].fillna(0.0)

In [27]:
tltrain_df['offnet_mou_6'].describe()

count    67231.000000
mean       198.874771
std        316.818355
min          0.000000
25%         34.860000
50%         96.480000
75%        232.990000
max       8362.360000
Name: offnet_mou_6, dtype: float64

In [28]:
# imputing the offnet_mou_6 column with 0.0
tltrain_df['offnet_mou_6'] = tltrain_df['offnet_mou_6'].fillna(0.0)

In [29]:
tltrain_df['offnet_mou_7'].describe()

count    67312.000000
mean       197.153383
std        322.482226
min          0.000000
25%         32.240000
50%         91.885000
75%        227.630000
max       7043.980000
Name: offnet_mou_7, dtype: float64

In [30]:
# imputing the offnet_mou_7 column with 0.0
tltrain_df['offnet_mou_7'] = tltrain_df['offnet_mou_7'].fillna(0.0)

In [31]:
tltrain_df['offnet_mou_8'].describe()

count    66296.000000
mean       196.543577
std        324.089234
min          0.000000
25%         31.575000
50%         91.800000
75%        229.345000
max      14007.340000
Name: offnet_mou_8, dtype: float64

In [32]:
# imputing the offnet_mou_8 column with 0.0
tltrain_df['offnet_mou_8'] = tltrain_df['offnet_mou_8'].fillna(0.0)

In [33]:
tltrain_df['roam_ic_mou_6'].describe()

count    67231.000000
mean         9.765435
std         57.374429
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max       2850.980000
Name: roam_ic_mou_6, dtype: float64

In [34]:
# imputing the roam_ic_mou_6 column with 0.0
tltrain_df['roam_ic_mou_6'] = tltrain_df['roam_ic_mou_6'].fillna(0.0)

In [35]:
tltrain_df['roam_ic_mou_7'].describe()

count    67312.000000
mean         7.014568
std         55.960985
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max       4155.830000
Name: roam_ic_mou_7, dtype: float64

In [36]:
# imputing the roam_ic_mou_7 column with 0.0
tltrain_df['roam_ic_mou_7'] = tltrain_df['roam_ic_mou_7'].fillna(0.0)

In [37]:
tltrain_df['roam_ic_mou_8'].describe()

count    66296.000000
mean         7.004892
std         53.408135
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max       4169.810000
Name: roam_ic_mou_8, dtype: float64

In [38]:
# imputing the roam_ic_mou_8 column with 0.0
tltrain_df['roam_ic_mou_8'] = tltrain_df['roam_ic_mou_8'].fillna(0.0)

In [39]:
# imputing the roam_og_mou_6 column with 0.0
tltrain_df['roam_og_mou_6'] = tltrain_df['roam_og_mou_6'].fillna(0.0)

In [40]:
# imputing the roam_og_mou_7 column with 0.0
tltrain_df['roam_og_mou_7'] = tltrain_df['roam_og_mou_7'].fillna(0.0)

In [41]:
# imputing the roam_og_mou_8 column with 0.0
tltrain_df['roam_og_mou_8'] = tltrain_df['roam_og_mou_8'].fillna(0.0)

In [42]:
tltrain_df['loc_og_t2t_mou_6'].describe()

count    67231.000000
mean        46.904854
std        150.971758
min          0.000000
25%          1.660000
50%         11.910000
75%         40.740000
max       6431.330000
Name: loc_og_t2t_mou_6, dtype: float64

In [43]:
# imputing the loc_og_t2t_mou_6 column with 0.0
tltrain_df['loc_og_t2t_mou_6'] = tltrain_df['loc_og_t2t_mou_6'].fillna(0.0)

In [44]:
# imputing the loc_og_t2t_mou_7 column with 0.0
tltrain_df['loc_og_t2t_mou_7'] = tltrain_df['loc_og_t2t_mou_7'].fillna(0.0)

In [45]:
# imputing the loc_og_t2t_mou_8 column with 0.0
tltrain_df['loc_og_t2t_mou_8'] = tltrain_df['loc_og_t2t_mou_8'].fillna(0.0)

In [47]:
columns = ['onnet_mou_6','onnet_mou_7','onnet_mou_8','offnet_mou_6','offnet_mou_7','offnet_mou_8','roam_ic_mou_6','roam_ic_mou_7','roam_ic_mou_8',
           'roam_og_mou_6','roam_og_mou_7','roam_og_mou_8','loc_og_t2t_mou_6','loc_og_t2t_mou_7','loc_og_t2t_mou_8','loc_og_t2m_mou_6','loc_og_t2m_mou_7',
           'loc_og_t2m_mou_8','loc_og_t2f_mou_6','loc_og_t2f_mou_7','loc_og_t2f_mou_8','loc_og_t2c_mou_6','loc_og_t2c_mou_7','loc_og_t2c_mou_8','loc_og_mou_6',
           'loc_og_mou_7','loc_og_mou_8','std_og_t2t_mou_6','std_og_t2t_mou_7','std_og_t2t_mou_8','std_og_t2m_mou_6','std_og_t2m_mou_7','std_og_t2m_mou_8','std_og_t2f_mou_6',
           'std_og_t2f_mou_7','std_og_t2f_mou_8','std_og_t2c_mou_6','std_og_t2c_mou_7','std_og_t2c_mou_8','std_og_mou_6','std_og_mou_7','std_og_mou_8','isd_og_mou_6',
           'isd_og_mou_7','isd_og_mou_8','spl_og_mou_6','spl_og_mou_7','spl_og_mou_8','og_others_6','og_others_7','og_others_8','loc_ic_t2t_mou_6','loc_ic_t2t_mou_7',
           'loc_ic_t2t_mou_8','loc_ic_t2m_mou_6','loc_ic_t2m_mou_7','loc_ic_t2m_mou_8','loc_ic_t2f_mou_6','loc_ic_t2f_mou_7','loc_ic_t2f_mou_8','loc_ic_mou_6','loc_ic_mou_7','loc_ic_mou_8',
           'std_ic_t2t_mou_6','std_ic_t2t_mou_7','std_ic_t2t_mou_8','std_ic_t2m_mou_6','std_ic_t2m_mou_7','std_ic_t2m_mou_8','std_ic_t2f_mou_6','std_ic_t2f_mou_7',
           'std_ic_t2f_mou_8','std_ic_t2o_mou_6','std_ic_t2o_mou_7','std_ic_t2o_mou_8','std_ic_mou_6','std_ic_mou_7','std_ic_mou_8','spl_ic_mou_6',
           'spl_ic_mou_7','spl_ic_mou_8','isd_ic_mou_6','isd_ic_mou_7','isd_ic_mou_8','ic_others_6','ic_others_7','ic_others_8']
for column in columns:
    tltrain_df[column] = tltrain_df[column].fillna(0.0)


len(tltrain_df.columns[tltrain_df.isna().any()].tolist())

38